File: 2024/haskell/day03b/Main.hs
module Main ( main ) where

import System.Environment ( getArgs, getProgName )
import System.Exit ( exitFailure )
import Text.Regex.TDFA ( (=~), AllTextMatches(getAllTextMatches) )

usage :: IO ()
usage = do
  progname <- getProgName
  putStrLn $ "usage: " ++ progname ++ " <file>"
  exitFailure

process :: String -> Int
process contents =
    let extractInsns str = getAllTextMatches (str =~ "mul\\([0-9]+,[0-9]+\\)|don't\\(\\)|do\\(\\)") :: [String]
        extractNumPair mulExpr = getAllTextMatches (mulExpr =~ "[0-9]+") :: [String]
        convertPair numPair = map read numPair :: [Int]
    in fst $ foldl
                (\(acc, enabled) insn ->
                  case insn of
                    "do()" -> (acc, True)
                    "don't()" -> (acc, False)
                    _ -> if enabled
                            then (acc + product (convertPair $ extractNumPair insn), enabled)
                            else (acc, enabled)
                )
                (0, True)
                (extractInsns contents)

main :: IO ()
main = do
  args <- getArgs
  case args of
    [filename] -> do
      contents <- readFile filename
      let result = process contents
      putStrLn $ "result = " ++ show result
    _ -> usage

File: 2024/rust/day03b/src/main.rs
use regex::Regex;
use std::env;
use std::fs;
use std::process;

fn usage() {
    let progname = env::args().next().unwrap();
    println!("usage: {progname} <file>");
    process::exit(1);
}

fn process(contents: &str) -> u32 {
    let mut result = 0;
    let mut enabled = true;
    let insn_re = Regex::new(r"(mul\([0-9]+,[0-9]+\)|don't|do)").unwrap();
    let mul_re = Regex::new(r"mul\(([0-9]+),([0-9]+)\)").unwrap();
    for (_, [insn]) in insn_re.captures_iter(contents).map(|c| c.extract()) {
        match insn {
            "don't" => enabled = false,
            "do" => enabled = true,
            _ => {
                if enabled {
                    for (_, [first_str, second_str]) in mul_re.captures_iter(insn).map(|c| c.extract()) {
                        let first = first_str.parse::<u32>().unwrap();
                        let second = second_str.parse::<u32>().unwrap();
                        result += first * second;
                    }
                }
            }
        }
    }
    result
}

fn main() {
    if env::args().count() < 2 {
        usage();
    }
    let filename = env::args().nth(1).unwrap();
    let contents = fs::read_to_string(filename).expect("read of input file failed");
    let result = process(&contents);
    println!("result = {result}");
}
